### ISSUE TITLE
Release version 35.0.0 is correct

### ISSUE THREAD

**Issue by tiran** [pyca/cryptography#6345]
Cryptography has adopted a new version scheme and went from **3.4.8** straight to release **35.0.0**. 

Checkout the [changelog](https://cryptography.io/en/latest/changelog/) and [new version scheme](https://cryptography.io/en/latest/api-stability/#versioning) for more information.

**Comment by renardeinside** [comment]
Hi @tiran 

This broke a dependency installation for pip-based setups in my CI pipelines, for instance:
```
cryptography 35.0.0 is installed but cryptography<4,>=0.6 is required by {'msal'}
```
What could be done to fix/avoud this on the installation side?

**Comment by graingert** [comment]
> Hi @tiran
> 
> This broke a dependency installation for pip-based setups in my CI pipelines, for instance:
> 
> ```
> cryptography 35.0.0 is installed but cryptography<4,>=0.6 is required by {'msal'}
> ```
> 
> What could be done to fix/avoid/workaround this on the installation side?

You will have to wait for them to remove the upper bound https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/413

**Comment by reaperhulk** [comment]
For those seeing this issue, please note that anyone pinning `<4` was assuming SemVer, which cryptography versions prior to `35.0` did not use. While our new versioning *is SemVer*, we also will increase the major version for every release, regardless of perceived compatibility or incompatibility.

**Comment by milo-minderbinder** [comment]
I agree that the old versioning scheme needed to change (if for no other reason than it was not strictly compatible with PEP 440), but there are a few things that are not clear to me, at least, with the new version scheme after reading the linked docs.
1. If a dependent package wants to specify that they want the latest version of `cryptography` that is compatible (no breaking changes) with `35.0.0`, they can specify `cryptography<38.0.0,>=35.0.0`, because any breaking changes or deprecation from `35.0.0` will be warned about for two major releases before they are actually introduced, correct?
2. Will security fixes be backported to previous major releases, and if so, for how many major release versions will security fixes be provided? 

Here's an example scenario I am concerned about:
* a package depends on `cryptography<38,>=35.0.0` because it wants the latest version of `cryptography` that is compatible (i.e. no breaking changes) with the current `35.0.0` release. 
* Then let's say that `cryptography` adds three new separate features across three new major releases in accordance with the new version scheme (so latest is now `38.0.0`).
* _Then_ let's say a new critical security issue is identified in `cryptography` which exists in all versions since, and including, `35.0.0`. 

Depending on whether/how security issues are backported to older major releases, the dependent package may not automatically get the security fix. And if security fixes _aren't_ backported, it's not obvious to the maintainers of the dependent package what the level of effort will be to update to the latest version of `cryptography` which has the security fix, because it's not immediately clear whether the three new major releases were just adding totally new features or if they introduced breaking changes to existing APIs.

With SemVer as it's _usually_ understood, this problem wouldn't be as much of an issue (in theory), because major version changes are reserved for breaking changes, whereas new features which don't break existing APIs or behavior can be provided in minor releases. Moreover, it is also theoretically easier to establish a policy for backporting security fixes in that model.

Maybe you don't foresee any of this as being a realistic issue because you're not planning to release new major versions often enough that it's unreasonable to expect dependent packages to release new versions every time `cryptography` has a major release, but if that's the case, it would be good to get an idea of how frequent major releases are expected to be generally.

**Comment by rayluo** [comment]
Hi @milo-minderbinder you seem to speak from the perspective of `cryptography`'s *downstream consumer*. I also maintain a package which depends on cryptography, so I can relate to your thoughts.

> I agree that the old versioning scheme needed to change (if for no other reason than it was not strictly compatible with PEP 440), but there are a few things that are not clear to me, at least, with the new version scheme after reading the linked docs.

I do not think PEP 440 was the motivation of this versioning scheme change.

PEP 440 i.e. the Python's own versioning scheme, and its actual practice (the entire history of 2.x series and 3.x series), are largely [SemVer](https://semver.org). But `cryptography` never uses SemVer before (see #5532). In a sense, `cryptography` deliberately chooses to use monotonic versioning. The breaking changes will therefore be picked up by downstream consumers who sets no upper bound, and then being forced to migrate/adopt their software to the latest incompatible `cryptography`.

The new scheme now is not really SemVer either, because downstream consumers can not adopt SemVer practice as-is. When a consumer is advised to not pin upper bound, we cannot say this is SemVer. (I have no problem of a software choosing non-SemVer. Just please do not advertise it as SemVer, then.)

> 
> 1. If a dependent package wants to specify that they want the latest version of `cryptography` that is compatible (no breaking changes) with `35.0.0`, they can specify `cryptography<38.0.0,>=35.0.0`, because any breaking changes or deprecation from `35.0.0` will be warned about for two major releases before they are actually introduced, correct?

Yes, if that dependent package wants to make sure their end users would *never* suddenly install a potentially incompatible cryptography in the future, this is the only option. When/if choosing this option, it would also imply that, the downstream packages' maintainers would need to roughly keep up with the release cadence of `cryptography`. After `cryptography 37.0.0` being released, we would need to smoke test it with our packages, and release our software with a new declaration `cryptography<40.0.0,>=35.0.0`. There is definitely more workload here for dependent package maintainers. Pick your poison.

> 
> 2. Will security fixes be backported to previous major releases, and if so, for how many major release versions will security fixes be provided?

Most likely, not. That was specifically the workload which the `cryptography` team wants to avoid. See it in details [here](https://github.com/pyca/cryptography/issues/5532#issuecomment-720252157). That is also why the monotonic versioning was chosen. Again, this is the trade-off between monotonic and SemVer.

> 
> 
> Here's an example scenario I am concerned about:
> 
> * a package depends on `cryptography<38,>=35.0.0` because it wants the latest version of `cryptography` that is compatible (i.e. no breaking changes) with the current `35.0.0` release.
> 
> * Then let's say that `cryptography` adds three new separate features across three new major releases in accordance with the new version scheme (so latest is now `38.0.0`).
> 
> * _Then_ let's say a new critical security issue is identified in `cryptography` which exists in all versions since, and including, `35.0.0`.
> 
> 
> Depending on whether/how security issues are backported to older major releases, the dependent package may not automatically get the security fix. And if security fixes _aren't_ backported, it's not obvious to the maintainers of the dependent package what the level of effort will be to update to the latest version of `cryptography` which has the security fix, because it's not immediately clear whether the three new major releases were just adding totally new features or if they introduced breaking changes to existing APIs.
> 
> With SemVer as it's _usually_ understood, this problem wouldn't be as much of an issue (in theory), because major version changes are reserved for breaking changes, whereas new features which don't break existing APIs or behavior can be provided in minor releases. Moreover, it is also theoretically easier to establish a policy for backporting security fixes in that model.
> 
> Maybe you don't foresee any of this as being a realistic issue because you're not planning to release new major versions often enough that it's unreasonable to expect dependent packages to release new versions every time `cryptography` has a major release, but if that's the case, it would be good to get an idea of how frequent major releases are expected to be generally.

Given the circumstances, the downstream package could choose to do that compatibility work, so that their end users do not have to. When a downstream library `Foo` is using SemVer and its `foo 1.2.3` depends on `cryptography<38.0.0,>=35.0.0`, and then later `foo 1.2.4` is released with updated dependency `cryptography<40.0.0,>=35.0.0`, the end users of `foo` will automatically obtain latest `foo 1.2.4` and `cryptography latest (up to 39)`, without being hit by potential breaking changes.
